File: test/ModularPipelines.UnitTests/KeyValueTests.cs
using ModularPipelines.Models;

namespace ModularPipelines.UnitTests;

public class KeyValueTests
{
    [Test]
    public async Task ImplicitOperator1()
    {
        KeyValue keyValue = ("one", "two");
        
        await using (Assert.Multiple())
        {
            await Assert.That(keyValue.Key).IsEqualTo("one");
            await Assert.That(keyValue.Value).IsEqualTo("two");
        }
    }

    [Test]
    public async Task ImplicitOperator2()
    {
        KeyValue keyValue = new Tuple<string, string>("one", "two");
        
        await using (Assert.Multiple())
        {
            await Assert.That(keyValue.Key).IsEqualTo("one");
            await Assert.That(keyValue.Value).IsEqualTo("two");
        }
    }

    [Test]
    public async Task ImplicitOperator3()
    {
        KeyValue keyValue = new KeyValuePair<string, string>("one", "two");
        
        await using (Assert.Multiple())
        {
            await Assert.That(keyValue.Key).IsEqualTo("one");
            await Assert.That(keyValue.Value).IsEqualTo("two");
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Models/KeyValueTests.fs
namespace ModularPipelines.UnitTests.FSharp.Models

open System
open System.Collections.Generic
open ModularPipelines.Models
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type KeyValueTests() =
    [<Test>]
    member _.ImplicitOperator1() = async {
        let keyValue: KeyValue = ("one", "two")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Key), "one"))
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Value), "two"))
    }

    [<Test>]
    member _.ImplicitOperator2() = async {
        let keyValue: KeyValue = Tuple<string, string>("one", "two")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Key), "one"))
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Value), "two"))
    }

    [<Test>]
    member _.ImplicitOperator3() = async {
        let keyValue: KeyValue = KeyValuePair<string, string>("one", "two")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Key), "one"))
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(keyValue.Value), "two"))
    }
